### ISSUE TITLE
json_serializer is not required for JSONB in SqlAlchemy, but json_deserializer is

### ISSUE THREAD

**Issue by nstewart** [cockroachdb/cockroach#23821]
Not familiar enough with SqlAlchemy to know if this is a bug, documentation issue, or enhancement. This took quite some time to figure out.

I created a simple object for the class for a `vehicles` table
```
class Vehicle(Base):
    __tablename__ = 'vehicles'
    id = Column(UUID, primary_key=True, default=generate_uuid)
    type = Column(String)
    city = Column(String)
    owner_id = Column(UUID)
    creation_time = Column(DateTime, default=datetime.datetime.now)
    status = Column(String)
    ext = Column(JSONB)
```

Adding a vehicle with a JSON document works fine:
`session.add(Vehicle(type=vehicle_type, city=random.choice(args.city), owner_id=str(uuid.uuid4()),
                        status=get_vehicle_availability(), ext=ext))`

Querying vehicles fails, however
`session.query(Vehicle).filter_by(status="available").limit(25).all()`

```
Traceback (most recent call last):
  File "./loadgen.py", line 183, in <module>
    action()
  File "./loadgen.py", line 144, in browse_vehicles
    session.query(Vehicle).filter_by(status="available").limit(25).all()
  File "/Library/Python/2.7/site-packages/sqlalchemy/orm/query.py", line 2737, in all
    return list(self)
  File "/Library/Python/2.7/site-packages/sqlalchemy/orm/loading.py", line 98, in instances
    util.raise_from_cause(err)
  File "/Library/Python/2.7/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Library/Python/2.7/site-packages/sqlalchemy/orm/loading.py", line 79, in instances
    rows = [proc(row) for row in fetch]
  File "/Library/Python/2.7/site-packages/sqlalchemy/orm/loading.py", line 493, in _instance
    loaded_instance, populate_existing, populators)
  File "/Library/Python/2.7/site-packages/sqlalchemy/orm/loading.py", line 593, in _populate_full
    dict_[key] = getter(row)
  File "/Library/Python/2.7/site-packages/sqlalchemy/sql/sqltypes.py", line 2155, in process
    return json_deserializer(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
TypeError: expected string or buffer
````

To fix this I had to add a custom deserializer
`engine = create_engine(args.conn_string, use_batch_mode=True, json_deserializer=jsonb_deserializer)`

even though it was a no op
```
def jsonb_deserializer(doc):
    return doc
```

Any idea what's going on there? I see that there is some sort of interaction with serializers at the dialect level according to [these docs](http://docs.sqlalchemy.org/en/rel_0_9/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSON)
 but it's not clear to me if that should apply to cockroachdb. Also note I installed the CockroachDB python package `pip install cockroachdb`

**Comment by justinj** [comment]
@nstewart can you try running this against Postgres and see if you have the same problem? I also don't think the CockroachDB python package is for applications - I think we use it for testing.

**Comment by justinj** [comment]
@nstewart or if you want to post your source somewhere I can do it.

**Comment by nstewart** [comment]
Sent you the source in a DM

**Comment by bdarnell** [comment]
>  I also don't think the CockroachDB python package is for applications - I think we use it for testing.

The CockroachDB python package is definitely intended for applications.

**Comment by justinj** [comment]
>The CockroachDB python package is definitely intended for applications.

My bad, I realized after investigating.

This needs to be fixed on the end of our Python adapter - we need to set `_has_native_json{,b}` on the CockroachDBDialect, but I'm not sure if there's a correct way to do that via deferring to the superclass or if we should just set it, so it's a slightly larger investigation. Since the deadline is looming for 2.0 fixes into Cockroach itself I'm going to postpone fixing this until after that.

**Comment by jseldess** [comment]
@justinj, do you think this warrants a callout in the sqlalchemy tutorial? If so, would you mind drafting something for me?

**Comment by justinj** [comment]
It won't warrant a callout once it's fixed, which I'll do at some point after we've completely closed the release-2.0 branch.
